ui/src/com/jediterm/terminal/ui/TerminalPanel.java

```diff
@@ -1221,13 +1239,82 @@ void setShape(CursorShape shape) {
     }
   }
 
+  public class BlinkingTextTracker {
+    private long myLastSlowTextBlink;
+    private long myLastRapidTextBlink;
+
+    private boolean mySlowTextBlinking;
+    private boolean myRapidTextBlinking;
+
+    public boolean isSlowTextBlinking() {
+      return mySlowTextBlinking;
+    }
+
+    public boolean isRapidTextBlinking() {
+      return myRapidTextBlinking;
+    }
+
+    public void changeStateIfNeeded() {
+      if (!isTextBlinkingEnabled())
+        return;
+      boolean repaint = false;
+      long currentTime = System.currentTimeMillis();
+
+      long slowRate = getSlowTextBlinkingPeriod();
+      long slowTriggerPoint = currentTime - slowRate;
+      if (myLastSlowTextBlink < slowTriggerPoint) {
+        mySlowTextBlinking = !mySlowTextBlinking;
+        if (myLastSlowTextBlink == 0)
+          myLastSlowTextBlink = currentTime;
+        else while (myLastSlowTextBlink < slowTriggerPoint)
+          myLastSlowTextBlink += slowRate;
+        repaint = true;
+      }
+
+      long rapidRate = getRapidTextBlinkingPeriod();
+      long rapidTriggerPoint = currentTime - rapidRate;
+      if (myLastRapidTextBlink < rapidTriggerPoint) {
+        myRapidTextBlinking = !myRapidTextBlinking;
+        if (myLastRapidTextBlink == 0)
+          myLastRapidTextBlink = currentTime;
+        else while (myLastRapidTextBlink < rapidTriggerPoint)
+          myLastRapidTextBlink += rapidRate;
+        repaint = true;
+      }
```

Comment:
Could you please try to get rid of the code duplication? Maybe have a dedicated class managing blinking of a particular period?

Reply:
Dedicated class added now, although I'm not sure if de-duplicating code here makes much sense considering that there would only ever be two type of blinking.